### ISSUE TITLE
[go_router] Preserve the state of routes

### ISSUE THREAD

**Issue by chunhtai** [flutter/flutter#99124]
copied from https://github.com/csells/go_router/issues/134

Currently there is no automatic way to have your routes remember their internal state as you navigate between them. This is often a desired UX feature as it is seen commonly on mobile and desktop platforms. For example, when I change tabs in a desktop app like Steam, it remembers my scroll position and selected sorting features (ephemeral view state).

The canonical way to do this in flutter is to use an Offstage widget, or something that wraps one, like an IndexedStack. The classic "nested navigator" pattern is a variation on this technique, where multiple stateful navigators are kept in memory using a TabBarView or IndexedStack.

I'm not sure if this is even possible with GoRouter, but would like to put this idea out there.

This is the core problem to solve. Given:
```dart
routes: [
  GoRoute(path: 'page1', pageBuilder: (_, state) => MaterialPage(key: state.pageKey, child: Page1())),
  GoRoute(path: 'page2', pageBuilder: (_, state) => MaterialPage(key: state.pageKey, child: Page2())),
]
```
When page1 is the path, the result of () => MaterialPage(key: state.pageKey, child: Page2()) is stashed Offstage somehow. Vice versa when `page2 is rendered. It wouldn't load everything at once, but rather lazy load and "remember" pages it had seen before. Likely there should be an option to opt-out certain views for persistence.

**Comment by Akaatsukhi** [comment]
Is there any workaround for this problem. is there any way through which state can be preserved maybe via key or anything. I am not sure how to do that. I would be glad if anyone can point out the doc or demo.

**Comment by arielbh** [comment]
@chunhtai @johnpryan 
Not sure how to upvote, but I would be happy if I could.
Perhaps if any of you can provide some clarity about it with 5.0 imminent (?)
Is it on the road map? and I just need to be more patient?
Should the community try to fix it? (perhaps even me? 😅 )

**Comment by chunhtai** [comment]
We are currently still in the planning phase. Given this is one of the higher priority issue, I think it will likely to be worked on first. If you would like to work on it, that will be great, too.

**Comment by esDotDev** [comment]
fwiw, when this was original discussed there did seem to be quite a bit of overlap here and the Restoration API used to restore stateful views when they are killed by the OS. 

It seems like a with a little bit of tweaking that API could be leveraged to give the impression of stateful routes, without actually needing to hold all the views offstage. 

Just food for thought, obviously would be great if things just persisted fully with no work from the developer. But it wouldn't be too bad if we have to use Restoration API, and then at least all our views will also work well when killed off by a mobile OS (or even an app crash) so you'd get two birds with one stone at least.

**Comment by tolo** [comment]
Instead of attempting to save and restore state, would it be an option to build a solution where a nested GoRouter object is used to preserve a navigation stack, while at the same time establishing a relationship with the parent GoRouter to facilitate deep linking etc? I may be wrong, but I think this is what a lot of people expected #99126 to accomplish. 

For instance, [Beamer](https://pub.dev/packages/beamer) does something like this. Here is an example with [bottom navigation](https://github.com/slovnicki/beamer/tree/master/examples/bottom_navigation_2). 

I don't know, maybe this has already been discussed and dismissed, if so I'm sorry for the noise.

**Comment by Dohmanlechx** [comment]
As of the new `go_router` version `5.0.0` I refactored our app from using `IndexedStack` to actually using the `ShellRoute` of `go_router` for our bottom navigation logic. It turned out that no state was preserved, which didn't make sense. When an app has a bottom navigation menu, we expect every tab to preserve its state while changing tabs.

Does someone have a solution for this?

**Comment by esDotDev** [comment]
My guess is that the scaffolding is in place for this now that there is proper nested Navigator support. I just thought I would bring up the RestorationAPI here, because it's a different way to tackle the problem that might co-exist with the ability to stash views in RAM.

**Comment by esDotDev** [comment]
fwiw, I created an issue on this here: https://github.com/flutter/flutter/issues/80303, it seems like we really are _so_ close to enabling general restoration with `StatefulWidgets` but it's not possible due to unfortunate design choices in the Restoration API (or I'm just missing it)

**Comment by tolo** [comment]
Yeah, I think I have stumbled over that issue before, and I agree that a state restoration based solution could possibly be used as a clean way of supporting "persistent" navigation stacks in for instance a BottomNavigationBar, without changing the intended way of configuring routes in go_router too much. I haven't used state restoration in Flutter myself that much though, but it seems like many parts that could form the foundation for such an implementation already exists in Flutter, even though this is perhaps a slightly different use case.

I can however imagine that in some cases (i.e. when using complex view hierarchies, logic and state, along with deeper navigation stacks etc), it would be easier and less complex to just keep the views and navigation stacks in memory. And in that case, a solution with separate, nested routers (like the Beamer example) might be preferable. Of course, nested GoRouters already work on a basic level, but support for things like deep linking are missing ([this demo/example](https://github.com/davidmartos96/go_router_bottom_nav_demo/) for instance is using nested GoRouters, without deep linking support).

But I guess the biggest question is how go_router could be modified to support these two different approaches. For the second approach at least, I imagine that it's more about documenting a particular usage pattern (with a good example), than making large changes in go_router.

**Comment by AceChen1** [comment]
@chunhtai hi hi any update about this part? we current face the same issue, A push to B, A,B will rebuild, B pop to A, A,B also will rebuild. This is strange...

**Comment by lucavenir** [comment]
I've seen the [pull request](https://github.com/flutter/packages/pull/2650) associated with this issue. That's phenomenal! I just wanted to ask: why is it introducing a new API for "stateful" shell navigation? Isn't that... "overcomplicating" things, a bit?

As underlined by several people in this thread, I would expect `ShellRoute` not to destroy previous states when navigating its subroutes. It would be weird (at least, for me) to define even more stateful logic just to let the router not destroy everything on a subroute change.

If anything, when changing routes entirely, I would demand a state management solution to handle caching and/or state (e.g. with riverpod).

Nonetheless, I just want to remark a big "thank you" for the team and the person working on the PR itself. It's extraordinary work.

**Comment by Junddao** [comment]
> @chunhtai hi hi any update about this part? we current face the same issue, (A push to B) => A,B will rebuild, (B pop to A) => A,B also will rebuild. This is strange...

@chunhtai  Is this issue fixable?
If not, you will need to modify your app to use the default navigator instead of go router.

I am always thank for your support.

**Comment by johnpryan** [comment]
@lucavenir the original idea was to add a parameter like `preserveState` to ShellRoute's constructor. The approach in https://github.com/flutter/packages/pull/2650 is a little bit different. Feel free to provide feedback on the API there - we're still in the process of reviewing it.

**Comment by esDotDev** [comment]
@johnpryan Does the `preserveState` solution allow all pages in the router to easily be preserved in memory? This is the key break-through with this approach. Multiple descendant navigators, with their pages always cached offstage by some ancestor widget. And it's recursive and can scale to any level/depth.

**Comment by johnpryan** [comment]
@esDotDev Since https://github.com/flutter/packages/pull/2650 provides a solution, I'm not sure if we will work on a design for `preserveState` right now.

**Comment by tolo** [comment]
Thanks @lucavenir, sorry for the late reply! 

I'm afraid introducing a new API is necessary, to be able to support multiple Navigators for a shell route, as well as the configuration related to which routes should belong to a certain Navigator etc. The goal is of course that the API should be a simple as possible, and with the help of feedback in the PR I think perhaps we've more or less reached a point where the API is good enough for a "StatefulShellRoute 1.0". But any feedback that moves the API in the right direction is of course always welcome. 😊

**Comment by leanhro2812** [comment]
Are there any new changes to this issue?

**Comment by rddewan** [comment]
any update on it ? please we required this feature

**Comment by sovetski** [comment]
It's really weird that there is no clean and especially official solution for this problem, which is still important and is even supported by native languages ​​like swift...

**Comment by lucavenir** [comment]
You can read updates and such in the linked [pull request](https://github.com/flutter/packages/pull/2650)

**Comment by ahmednfwela** [comment]
@chunhtai this issue can finally get closed? :)

**Comment by johnpryan** [comment]
Yes, I think we can close this now. Pull request https://github.com/flutter/packages/pull/2650 is merged and published as version 7.1.0
